packages/mermaid/scripts/create-types-from-json-schema.mjs

```diff
@@ -0,0 +1,250 @@
+    if (schema['allOf']) {
+      const { allOf, ...schemaWithoutAllOf } = schema;
+      return {
+        ...schemaWithoutAllOf,
+        extends: allOf,
+      };
+    }
+    return schema;
+  }
+
+  /**
+   * For backwards compatibility with older Mermaid Typescript defs,
+   * we need to make all value optional instead of required.
+   *
+   * This is because the `MermaidConfig` type is used as an input, and everything is optional,
+   * since all the required values have default values.s
+   *
+   * In the future, we should make make the input to Mermaid `Partial<MermaidConfig>`.
+   *
+   * @todo TODO: Remove this function when Mermaid releases a new breaking change.
+   * @param {JSONSchemaType} schema - The input schema.
+   * @returns {JSONSchemaType} The schema with all required values removed.
+   */
+  function removeRequired(schema) {
+    return { ...schema, required: [] };
+  }
+
+  /**
+   * This is a temporary hack to control the order the types are generated in.
+   *
+   * By default, json-schema-to-typescript outputs the $defs in the order they
+   * are used, then any unused schemas at the end.
+   *
+   * **The only purpose of this function is to make the `git diff` simpler**
+   * **We should remove this later to simplify the code**
+   *
+   * @todo TODO: Remove this function in a future PR.
+   * @param {JSONSchemaType} schema - The input schema.
+   * @returns {JSONSchemaType} The schema with all `$ref`s removed.
+   */
+  function unrefSubschemas(schema) {
+    return {
+      ...schema,
+      properties: Object.fromEntries(
+        Object.entries(schema.properties).map(([key, propertySchema]) => {
+          if (MERMAID_CONFIG_DIAGRAM_KEYS.includes(key)) {
+            const { $ref, ...propertySchemaWithoutRef } = propertySchema;
+            const [
+              _root, // eslint-disable-line @typescript-eslint/no-unused-vars
+              _defs, // eslint-disable-line @typescript-eslint/no-unused-vars
+              defName,
+            ] = $ref.split('/');
+            return [
+              key,
+              {
+                ...propertySchemaWithoutRef,
+                tsType: defName,
+              },
+            ];
+          }
+          return [key, propertySchema];
+        })
+      ),
+    };
+  }
+
+  assert.ok(mermaidConfigSchema.$defs);
+  const modifiedSchema = {
+    ...unrefSubschemas(removeRequired(mermaidConfigSchema)),
+
+    $defs: Object.fromEntries(
+      Object.entries(mermaidConfigSchema.$defs).map(([key, subSchema]) => {
+        // every field in every diagram config is optional, **EXCEPT**
+        // QuadrantChartConfig, where every field is required in TypeScript
+        // see https://github.com/mermaid-js/mermaid/blame/9c011abbd424641172b696c038490a36ec99c518/packages/mermaid/src/config.type.ts#L231-L250
+        if (key === 'QuadrantChartConfig') {
+          return [key, replaceAllOfWithExtends(subSchema)]; // keep all required fields
+        }
```

Comment:
I think this is a mistake in the `QuadrantChartConfig`. 
All the configs should be optional, with the default being specified in the defaultConfig file.

Reply:
I've made this fix in a separate PR: https://github.com/mermaid-js/mermaid/pull/4602, so that this PR will change nothing from a user perspective, all default config values and all TypeScript types will be exactly the same.

I've rebased my changes in this PR to add https://github.com/mermaid-js/mermaid/pull/4602/commits/38013de7114966e8b1a83396703ef8158bb34466 and removed these lines!